bigquery/google/cloud/bigquery/magics.py

```diff
@@ -0,0 +1,250 @@
+        amount of time for the query to complete will not be cleared after the
+        query is finished. By default, this information will be displayed but
+        will be cleared after the query is finished.
+    * ``<query>`` (required, cell argument):
+        SQL query to run.
+
+    Returns:
+        A :class:`pandas.DataFrame` with the query results.
+
+    .. note::
+        All queries run using this magic will run using the context client
+        (:func:`Context.client <google.cloud.bigquery.magics.Context.client>`)
+
+    Examples:
+        The following examples can be run in an IPython notebook after loading
+        the bigquery IPython extension (see ``In[1]``) and setting up
+        Application Default Credentials.
+
+    .. code-block:: python
+
+        In [1]: %load_ext google.cloud.bigquery
+
+        In [2]: %%bigquery
+           ...: SELECT name, SUM(number) as count
+           ...: FROM `bigquery-public-data.usa_names.usa_1910_current`
+           ...: GROUP BY name
+           ...: ORDER BY count DESC
+           ...: LIMIT 3
+
+        Out[2]:       name    count
+           ...: -------------------
+           ...: 0    James  4987296
+           ...: 1     John  4866302
+           ...: 2   Robert  4738204
+
+        In [3]: %%bigquery df --project my-alternate-project --verbose
+           ...: SELECT name, SUM(number) as count
+           ...: FROM `bigquery-public-data.usa_names.usa_1910_current`
+           ...: WHERE gender = 'F'
+           ...: GROUP BY name
+           ...: ORDER BY count DESC
+           ...: LIMIT 3
+        Executing query with job ID: bf633912-af2c-4780-b568-5d868058632b
+        Query executing: 2.61s
+        Query complete after 2.92s
+
+        Out[3]:          name    count
+           ...: ----------------------
+           ...: 0        Mary  3736239
+           ...: 1    Patricia  1568495
+           ...: 2   Elizabeth  1519946
+
+        In [4]: df
+
+        Out[4]:          name    count
+           ...: ----------------------
+           ...: 0        Mary  3736239
+           ...: 1    Patricia  1568495
+           ...: 2   Elizabeth  1519946
+
+"""
+
+from __future__ import print_function
+
+import time
+from concurrent import futures
+
+try:
+    import IPython
+    from IPython import display
+    from IPython.core import magic_arguments
+except ImportError:  # pragma: NO COVER
+    raise ImportError('This module can only be loaded in IPython.')
+
+from google.cloud.bigquery.client import Client
+from google.cloud.bigquery.job import QueryJobConfig
+
+
+class Context(object):
+    """Storage for objects to be used throughout an IPython notebook session
+
+    A Context object is initialized when the ``magics`` module is imported,
+    and can be found at ``google.cloud.bigquery.magics.context``.
+    """
+    def __init__(self):
+        self._client = None
+
+    @property
+    def client(self):
```

Comment:
As discussed in another thread with @craigcitro , since `Client` requires a project, but not all forms of auth provide a default project (user auth doesn't), this should be a `Credentials` object, instead.

That also works around the issue where project can't be overriden in query. By cacheing Credentials, that means we create a new Client object every time in the query magic.

Reply:
Updated to store credentials instead of client in e577864